RecoHI/HiJetAlgos/plugins/HiFJRhoFlowModulationProducer.cc

```diff
@@ -48,40 +92,78 @@ class HiFJRhoFlowModulationProducer : public edm::stream::EDProducer<> {
       doFreePlaneFit_(iConfig.getParameter<bool>("doFreePlaneFit")),
       doJettyExclusion_(iConfig.getParameter<bool>("doJettyExclusion")),
       evtPlaneLevel_(iConfig.getParameter<int>("evtPlaneLevel")),
+      pfCandidateEtaCut_(iConfig.getParameter<double>("pfCandidateEtaCut")),
+      pfCandidateMinPtCut_(iConfig.getParameter<double>("pfCandidateMinPtCut")),
+      pfCandidateMaxPtCut_(iConfig.getParameter<double>("pfCandidateMaxPtCut")),
+      firstFittedVn_(iConfig.getParameter<int>("firstFittedVn")),
+      lastFittedVn_(iConfig.getParameter<int>("lastFittedVn")),
       jetToken_(doJettyExclusion_ ? consumes<reco::JetView>(iConfig.getParameter<edm::InputTag>("jetTag"))
                                   : edm::EDGetTokenT<reco::JetView>()),
-      pfCandsToken_(consumes<reco::PFCandidateCollection>(iConfig.getParameter<edm::InputTag>("pfCandSource"))),
+      pfCandidateToken_(consumes<pat::PackedCandidateCollection>(iConfig.getParameter<edm::InputTag>("pfCandSource"))),
       evtPlaneToken_(consumes<reco::EvtPlaneCollection>(iConfig.getParameter<edm::InputTag>("EvtPlane"))) {
   produces<std::vector<double>>("rhoFlowFitParams");
+
+  // Converter to get PF candidate ID from packed candidates
+  converter_ = reco::PFCandidate();
+
+  // Sanity check for the fitted flow component input
+  if (firstFittedVn_ < 1)
+    firstFittedVn_ = 2;
+  if (lastFittedVn_ < 1)
+    lastFittedVn_ = 2;
+  if (firstFittedVn_ > lastFittedVn_) {
+    int flipper = lastFittedVn_;
+    lastFittedVn_ = firstFittedVn_;
+    firstFittedVn_ = flipper;
+  }
+
+  // Calculate the number of flow components
+  const int nFlow = lastFittedVn_ - firstFittedVn_ + 1;
+
+  // Define all fit and weight functions
   TMinuitMinimizer::UseStaticMinuit(false);
-  lineFit_p_ = std::make_unique<TF1>("lineFit", lineFunction, -TMath::Pi(), TMath::Pi());
-  flowFit_p_ = std::make_unique<TF1>("flowFit", flowFunction, -TMath::Pi(), TMath::Pi());
+  flowFit_p_ = std::make_unique<TF1>("flowFit", flowFunction, -TMath::Pi(), TMath::Pi(), nFlow * 2 + 3);
+  flowFit_p_->FixParameter(0, nFlow);           // The first parameter defines the number of fitted flow components
+  flowFit_p_->FixParameter(1, firstFittedVn_);  // The second parameter defines the first fitted flow component
+  pfWeightFunction_ = std::make_unique<TF2>("weightFunction", weightFunction, 0, TMath::Pi(), -5, 5, 1);
+  pfWeightFunction_->SetParameter(0, pfCandidateEtaCut_);  // Set the allowed eta range for particle flow candidates
 }
 
 // ------------ method called to produce the data  ------------
 void HiFJRhoFlowModulationProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
-  int nPhiBins = 10;
-  auto const& pfCands = iEvent.get(pfCandsToken_);
+  // Get the particle flow candidate collection
+  auto const& pfCands = iEvent.get(pfCandidateToken_);
+
+  // If we are reading the event plane information for the forest, find the event planes
   static constexpr int kMaxEvtPlane = 29;
```

Comment:
where does this value of 29 come from?

Reply:
That used to be the number of different event planes reconstructed in the 2018 PbPb AOD reconstruction. It looks like this is updated since. It might be better to use hi::NumEPNames, since this is a constant that tracks the number of event planes in the current reconstruction.